optuna/pruners/_wilcoxon.py

```diff
@@ -0,0 +1,189 @@
+
+    .. seealso::
+        Please refer to :meth:`~optuna.trial.Trial.report`.
+
+    Example:
+
+        .. testcode::
+
+            import optuna
+            import numpy as np
+
+
+            # For demonstrative purposes, we will use a toy evaluation function.
+            # We will minimize the mean value of `eval_func` over the input dataset.
+            def eval_func(param, input_):
+                return (param - input_) ** 2
+
+
+            input_data = np.linspace(-1, 1, 100)
+
+            # It is recommended to shuffle the input data once before optimization.
+            np.random.shuffle(input_data)
+
+
+            def objective(trial):
+                s = 0.0
+                for i in range(len(input_data)):
+                    param = trial.suggest_uniform("param", -1, 1)
+                    s += eval_func(param, input_data[i])
+
+                    trial.report(s / (i + 1), i + 1)
+                    if trial.should_prune():
+                        raise optuna.TrialPruned()
+
+                return s / len(input_data)
+
+
+            study = optuna.study.create_study(
+                pruner=optuna.pruners.WilcoxonPruner(p_threshold=0.1)
+            )
+            study.optimize(objective, n_trials=100)
+
+    Args:
+        p_threshold:
+            The p-value threshold for pruning. This value should be between 0 and 1.
+            A trial will be pruned whenever the pruner is sure up to the given p-value
+            that the current trial is worse than the best trial.
+            The larger this value is, the more aggressive pruning will be performed.
+            Defaults to 0.1.
+
+            .. note::
+                Contrary to the usual statistical wisdom, this pruner repeatedly
+                performs statistical tests between the current trial and the
+                current best trial with increasing samples.
+                Please expect around ~2x probability of falsely pruning
+                a good trial, compared to the usual false positive rate of
+                performing the statistical test only once.
+
+        n_startup_steps:
+            The number of steps before which no trials are pruned.
+            Defaults to 0 (pruning kicks in from the very first step).
+    """  # NOQA: E501
+
+    def __init__(
+        self,
+        *,
+        p_threshold: float = 0.1,
+        n_startup_steps: int = 0,
+    ) -> None:
+        if n_startup_steps < 0:
+            raise ValueError(f"n_startup_steps must be nonnegative but got {n_startup_steps}.")
+        if not 0.0 <= p_threshold <= 1.0:
+            raise ValueError(f"p_threshold must be between 0 and 1 but got {p_threshold}.")
+
+        self._n_startup_steps = n_startup_steps
+        self._p_threshold = p_threshold
+
+    def prune(self, study: "optuna.study.Study", trial: FrozenTrial) -> bool:
+        ss = _LazyImport("scipy.stats")
```

Comment:
`try_import` would import scipy regardless of whether you use `WilcoxonPruner`. This would slow down `import optuna`. `_LazyImport` does that lazily, so you import `scipy` only when you use `scipy`. (`ImportError` is emitted at that time, too.)

Reply:
```suggestion
        import scipy.stats as ss
```